security/advancedtls/advancedtls_integration_test.go

```diff
@@ -0,0 +1,509 @@
+		}
+	} else {
+		conn, err = grpc.DialContext(connCtx, address, grpc.WithTransportCredentials(creds), grpc.WithBlock())
+		if err != nil {
+			return nil, nil, fmt.Errorf("client failed to connect to %s. Error: %v", address, err)
+		}
+	}
+	greetClient := pb.NewGreeterClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	_, err = greetClient.SayHello(ctx, &pb.HelloRequest{Name: msg})
+	if want, got := shouldFail == true, err != nil; got != want {
+		return nil, nil, fmt.Errorf("want and got mismatch,  want shouldFail=%v, got fail=%v, rpc error: %v", want, got, err)
+	}
+	return conn, greetClient, nil
+}
+
+// Test Scenarios:
+// At initialization(stage = 0), client will be initialized with cert clientPeer1 and clientTrust1, server with serverPeer1 and serverTrust1.
+// The mutual authentication works at the beginning, since clientPeer1 is trusted by serverTrust1, and serverPeer1 by clientTrust1.
+// At stage 1, client changes clientPeer1 to clientPeer2. Since clientPeer2 is not trusted by serverTrust1, following rpc calls are expected
+// to fail, while the previous rpc calls are still good because those are already authenticated.
+// At stage 2, the server changes serverTrust1 to serverTrust2, and we should see it again accepts the connection, since clientPeer2 is trusted
+// by serverTrust2.
+func TestClientPeerCertReloadServerTrustCertReload(t *testing.T) {
+	stage := &stageInfo{}
+	cs := &certStore{}
+	err := cs.loadCerts()
+	if err != nil {
+		t.Fatalf("load certs failed, err: %v", err)
+	}
+	verifyFunc := func(params *VerificationFuncParams) (*VerificationResults, error) {
+		return &VerificationResults{}, nil
+	}
+	getCertForClient := func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		switch stage.read() {
+		case 0:
+			return &cs.clientPeer1, nil
+		default:
+			return &cs.clientPeer2, nil
+		}
+	}
+	getRootCAsForServer := func(params *GetRootCAsParams) (*GetRootCAsResults, error) {
+		switch stage.read() {
+		case 0, 1:
+			return &GetRootCAsResults{TrustCerts: cs.serverTrust1}, nil
+		default:
+			return &GetRootCAsResults{TrustCerts: cs.serverTrust2}, nil
+		}
+	}
+	// Start a server using ServerOptions in another goroutine.
+	serverOptions := &ServerOptions{
+		Certificates:   []tls.Certificate{cs.serverPeer1},
+		GetCertificate: nil,
+		RootCertificateOptions: RootCertificateOptions{
+			RootCACerts: nil,
+			GetRootCAs:  getRootCAsForServer,
+		},
+		RequireClientCert: true,
+	}
+	serverTLSCreds, err := NewServerCreds(serverOptions)
+	s := grpc.NewServer(grpc.Creds(serverTLSCreds))
+	defer s.Stop()
+	go func(s *grpc.Server) {
+		lis, err := net.Listen("tcp", port)
+		// defer lis.Close()
+		if err != nil {
+			t.Fatalf("Failed to listen: %v", err)
+		}
+		pb.RegisterGreeterServer(s, &serverImpl{})
+		if err := s.Serve(lis); err != nil {
+			log.Fatalf("failed to serve: %v", err)
+		}
+	}(s)
+
```

Comment:
I would like to suggest some refactoring for the tests. The driving factor is that although the stages idea works, it is a bit confusing to follow what is happening in the code.

How about we add table-driven tests here where each case takes exactly what it needs to run. You probably need to create `ClientConn` objects before the loop and pass them to the various test cases. If such an object is passed, then `callAndVerifyInSameRPC` is called, otherwise, `callAndVerifyWithNewRPC` is called.

The same applies for the other tests below.

Reply:
After thinking of this for a while, I realized the tests could also be table-driven by combining the three test cases into one. We now have only one test `TestEnd2End` containing three test cases, represented as the table elements. 
This might differ a little bit from the way you suggested, but it looks good to me overall. Could you please take a look and see if we could improve based on the updated code? I am happy to discuss further :-)